File: politiks-api/src/main/java/org/politiks/model/dto/CandidatoDTO.java
package org.politiks.model.dto;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;

/**
 * DTO para transferência de dados de Candidato entre a API e o cliente.
 * Segue o padrão DTO para desacoplar a API da camada de domínio.
 */
public class CandidatoDTO {
    
    private Long id;
    
    @NotBlank(message = "O nome do candidato é obrigatório")
    @Size(min = 2, max = 100, message = "O nome do candidato deve ter entre 3 e 100 caracteres")
    private String nome;
    
    @NotBlank(message = "O CPF do candidato é obrigatório")
    @Size(min = 11, max = 14, message = "O CPF deve ter entre 11 e 14 caracteres")
    private String cpf;
    
    private String dataNascimento;
    
    private String biografia;
    
    private String cargoPleiteado;
    
    @NotNull(message = "O ID do partido é obrigatório")
    private Long partidoId;
    
    private String partidoNome;
    
    private String partidoSigla;
    
    // Construtores
    public CandidatoDTO() {
    }
    
    public CandidatoDTO(Long id, String nome, String cpf, String dataNascimento, 
                      String biografia, String cargoPleiteado, Long partidoId) {
        this.id = id;
        this.nome = nome;
        this.cpf = cpf;
        this.dataNascimento = dataNascimento;
        this.biografia = biografia;
        this.cargoPleiteado = cargoPleiteado;
        this.partidoId = partidoId;
    }
    
    public CandidatoDTO(Long id, String nome, String cpf, String dataNascimento, 
                      String biografia, String cargoPleiteado, Long partidoId, 
                      String partidoNome, String partidoSigla) {
        this.id = id;
        this.nome = nome;
        this.cpf = cpf;
        this.dataNascimento = dataNascimento;
        this.biografia = biografia;
        this.cargoPleiteado = cargoPleiteado;
        this.partidoId = partidoId;
        this.partidoNome = partidoNome;
        this.partidoSigla = partidoSigla;
    }
    
    // Getters e Setters
    public Long getId() {
        return id;
    }
    
    public void setId(Long id) {
        this.id = id;
    }
    
    public String getNome() {
        return nome;
    }
    
    public void setNome(String nome) {
        this.nome = nome;
    }
    
    public String getCpf() {
        return cpf;
    }
    
    public void setCpf(String cpf) {
        this.cpf = cpf;
    }
    
    public String getDataNascimento() {
        return dataNascimento;
    }
    
    public void setDataNascimento(String dataNascimento) {
        this.dataNascimento = dataNascimento;
    }
    
    public String getBiografia() {
        return biografia;
    }
    
    public void setBiografia(String biografia) {
        this.biografia = biografia;
    }
    
    public String getCargoPleiteado() {
        return cargoPleiteado;
    }
    
    public void setCargoPleiteado(String cargoPleiteado) {
        this.cargoPleiteado = cargoPleiteado;
    }
    
    public Long getPartidoId() {
        return partidoId;
    }
    
    public void setPartidoId(Long partidoId) {
        this.partidoId = partidoId;
    }
    
    public String getPartidoNome() {
        return partidoNome;
    }
    
    public void setPartidoNome(String partidoNome) {
        this.partidoNome = partidoNome;
    }
    
    public String getPartidoSigla() {
        return partidoSigla;
    }
    
    public void setPartidoSigla(String partidoSigla) {
        this.partidoSigla = partidoSigla;
    }
}

File: politiks-bff-nest/src/dtos/candidato.dto.ts
import { IsNotEmpty, IsString, IsOptional, IsNumber, MinLength, MaxLength } from 'class-validator';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { PartidoDto } from './partido.dto';

/**
 * DTO para transferência de dados de Candidato.
 * Segue o padrão DTO (Data Transfer Object) para desacoplar a API da camada de domínio.
 */
export class CandidatoDto {
  @ApiPropertyOptional({ description: 'ID do candidato', example: 1 })
  @IsOptional()
  @IsNumber()
  id?: number;

  @ApiProperty({ description: 'Nome do candidato', example: 'João Silva', minLength: 2, maxLength: 100 })
  @IsNotEmpty({ message: 'O nome do candidato é obrigatório' })
  @IsString({ message: 'O nome deve ser uma string' })
  @MinLength(2, { message: 'O nome do candidato deve ter pelo menos 2 caracteres' })
  @MaxLength(100, { message: 'O nome do candidato deve ter no máximo 100 caracteres' })
  nome: string;

  @ApiProperty({ description: 'CPF do candidato', example: '123.456.789-00', minLength: 11, maxLength: 14 })
  @IsNotEmpty({ message: 'O CPF do candidato é obrigatório' })
  @IsString({ message: 'O CPF deve ser uma string' })
  @MinLength(11, { message: 'O CPF deve ter pelo menos 11 caracteres' })
  @MaxLength(14, { message: 'O CPF deve ter no máximo 14 caracteres' })
  cpf: string;

  @ApiPropertyOptional({ description: 'Data de nascimento do candidato', example: '1980-01-01' })
  @IsOptional()
  @IsString({ message: 'A data de nascimento deve ser uma string' })
  dataNascimento?: string;

  @ApiPropertyOptional({ description: 'Biografia do candidato', example: 'Político com experiência em...' })
  @IsOptional()
  @IsString({ message: 'A biografia deve ser uma string' })
  biografia?: string;

  @ApiPropertyOptional({ description: 'Cargo pleiteado pelo candidato', example: 'Vereador' })
  @IsOptional()
  @IsString({ message: 'O cargo pleiteado deve ser uma string' })
  cargoPleiteado?: string;

  @ApiProperty({ description: 'ID do partido ao qual o candidato pertence', example: 1 })
  @IsNotEmpty({ message: 'O ID do partido é obrigatório' })
  @IsNumber({}, { message: 'O ID do partido deve ser um número' })
  partidoId: number;

  @ApiPropertyOptional({ description: 'Nome do partido', example: 'Partido Exemplo' })
  @IsOptional()
  @IsString({ message: 'O nome do partido deve ser uma string' })
  partidoNome?: string;

  @ApiPropertyOptional({ description: 'Sigla do partido', example: 'PEX' })
  @IsOptional()
  @IsString({ message: 'A sigla do partido deve ser uma string' })
  partidoSigla?: string;
}

/**
 * DTO para resposta de Candidato com dados completos do partido.
 * Estende o DTO básico de Candidato.
 */
export class CandidatoComPartidoDto extends CandidatoDto {
  @ApiPropertyOptional({ 
    description: 'Dados completos do partido',
    type: PartidoDto
  })
  partido?: PartidoDto;
}
